Implement change() to properly reflect the correct UTC time

seanprashad opened the pull request:
### Summary

Fixes #32550 - implemented `change()` to properly convert the value to UTC on the date component (_which may be different to the conversion_) to UTC on `2000-01-01`.

### Other Information

N/A

Commit:
Implement change() to convert to "2001-01-01" first

Commit:
Normalize the date component to account for DST

SeanPrashad commented:
I went ahead and updated the respective test (inside [activerecord/test/cases/quoting_test.rb](https://github.com/rails/rails/blob/master/activerecord/test/cases/quoting_test.rb#L70)) for this patch - should I go ahead and add a new, separate test?

pixeltrix commented on activerecord/lib/active_record/connection_adapters/abstract/quoting.rb:
```diff
@@ -130,6 +130,7 @@ def quoted_date(value)
 
       def quoted_time(value) # :nodoc:
+        value = value.change(year: 2000, month: 1, day: 1)
         quoted_date(value).sub(/\A\d\d\d\d-\d\d-\d\d /, "")
```
If we know the date is 2001-01-01 we can probably just strip off the first 11 characters and save ourselves the `sub`

pixeltrix reviewed (changes requested):
We need to add a test that reflects the scenario in the issue where the offset changes because the current date is DST but 2000-01-01 is not. Also would need to have a CHANGELOG entry and the commit description isn't really accurate - it's more about normalising the date component to the same date than anything to do with UTC.

SeanPrashad commented:
Thanks for the feedback @pixeltrix - I've replaced `sub` with `splice` to get the rest of the string after 11 characters and added a `CHANGELOG` entry.

I've having some trouble with identifying how the test should look like. From what I understand, I should be aiming for something like this:

```
def test_quoted_time_dst
  with_timezone_config default: :utc do
    # Create a DateTime object with a timezone that is currently in DST
    expected = ...

    expected = expected.getutc.to_s(:db).sub("2000-01-01 ", "")

    assert_equal expected, @quoter.quoted_time(t)
  end
end
```

Are there any resources/tests that I can investigate to help get us across this final hurdle?

rafaelfranca commented on activerecord/CHANGELOG.md:
```diff
@@ -23,5 +23,9 @@
 
     *DHH*
 
+*   Normalize the date component to account for DST.
```
New changelog entires are on the top of the file

SeanPrashad replied:
FIxed - thanks!

pixeltrix commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+*   Normalize the date component to account for DST.
```
Can you changes this to:

```
*   Normalize the date component for `time` columns to the epoch date of 2000-01-01.
```

pixeltrix commented on activerecord/test/cases/quoting_test.rb:
```diff
@@ -71,8 +71,19 @@ def test_quoted_time_utc
+          t = Time.now.change(usec: 0)
+
+          expected = t.change(year: 2000, month: 1, day: 1)
+          expected = expected.getutc.to_s(:db).sub("2000-01-01 ", "")
```
Can you change this test to check that a time that is currently in DST is not in DST when quoted. You'll need to change `with_timezone_config` to use a zone other than `:utc` and use a fixed date/time rather than relying on `Time.now`. You'll also need to test this for when the Active Record timestamp format is both `:utc` and `:local`.

Commit:
Update CHANGELOG message

Commit:
Add tests to check normalization of time columns

SeanPrashad commented on activerecord/test/cases/quoting_test.rb:
```diff
@@ -71,9 +71,33 @@ def test_quoted_time_utc
+
+      def test_quoted_time_utc_dst
+        with_timezone_config default: :utc do
+          t = Time.new(2000, 7, 1)
```
This date was chosen as per the examples in Ruby's documentation for [`.dst?`](https://ruby-doc.org/core-2.2.0/Time.html#method-i-dst-3F)

SeanPrashad commented:
I've taken a crack at the two tests as per your recommendations @pixeltrix. I think we're close and would greatly appreciate your feedback!

bogdanvlviv commented:
This can be closed since #33358 was merged.
